src/main.rs

```diff
@@ -1,3 +1,192 @@
-fn main() {
-    println!("Hello, world!");
+extern crate futures;
+extern crate reqwest;
+extern crate tokio;
+
+use futures::Future;
+use reqwest::r#async::Client;
+
+trait InfluxDbQuery {
+    fn build<'a>(self) -> String;
 }
+
+impl InfluxDbQuery {
+    pub fn write() -> InfluxDbWrite {
+        InfluxDbWrite {
+            _measurement: String::from("marina_3"),
+            _fields: Vec::new(),
+            _tags: Vec::new(),
+        }
+    }
+
+    // pub fn read() {}
+}
+
+pub struct InfluxDbWrite {
+    _fields: Vec<(String, String)>,
+    _tags: Vec<(String, String)>,
+    _measurement: String,
+    // precision: Precision
+}
+
+impl InfluxDbWrite {
+    fn add_field<'a, S>(&'a mut self, point: S, value: S) -> &'a mut Self
+    where
+        S: Into<String>,
+    {
+        self._fields.push((point.into(), value.into()));
+        self
+    }
+
+    fn add_tag<'a, S>(&'a mut self, tag: S, value: S) -> &'a mut Self
+    where
+        S: Into<String>,
+    {
+        self._tags.push((tag.into(), value.into()));
+        self
+    }
+}
+
+impl InfluxDbQuery for InfluxDbWrite {
+    // fixme: time (with precision) and measurement
+    fn build<'a>(self) -> String {
+        let tags = self
+            ._tags
+            .into_iter()
+            .map(|(tag, value)| format!("{tag}={value}", tag = tag, value = value))
+            .collect::<Vec<String>>()
+            .join(",");
```

Comment:
Collecting into a `Vec` only to join to a `String` is expensive. Use [`join`](https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.join) on iterators instead.

Reply:
Good catch!